### PR TITLE
Extracted area type into a reference table, improved error handling

### PR DESCRIPTION
**Pull request by YetAnotherMatt**
### What

Extracted area type into a separate reference table, improved error handling. Still need to sort out what to do about classifications having multiple hierarchies with the same code in a single file (potentially merge them into a single hierarchy).

### How to review

Run tests, build, run:
dp-dd-hierarchy-importer -type=g http://web.ons.gov.uk/ons/api/data/hierarchies/hierarchy/2011STATH.json?apikey=API_KEY&levels=1,2,3,4,5,6,7,8

### Who can review

@fenallen, or anyone.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -51,14 +47,20 @@ Concepts are hierarchical, having a parent, level, order and collection count
 - Socio-Economic: none
 - Social: http://web.ons.gov.uk/ons/api/data/classifications.xml?apikey=API_KEY&context=Social
 
-Classifications can be hierarchical, having a parent, order, IsTotal and SubTotal flags. Some classification hierarchies can be flat.
+Classifications can be hierarchical, having a parent, order, IsTotal and SubTotal flags. 
+Some classification hierarchies can be flat, and we will only import classifications with a deep tree (such as COICOP: http://web.ons.gov.uk/ons/api/data/classification/CL_0000641.xml?apikey=API_KEY&context=Economic).
```
Comment (NeilMadden) [inline README.md]: Not sure I follow how a hierarchy can be "flat"? Maybe reword to something like "Some classifications can be flat (non-hierarchical)..."?

File: geography/loader.go [hunk 2]
```diff
@@ -5,8 +5,11 @@ import (
 
 	"github.com/ONSdigital/dp-dd-hierarchy-importer/parser"
 	. "github.com/ONSdigital/dp-dd-hierarchy-importer/sql"
+	"fmt"
```
Comment (NeilMadden) [inline geography/loader.go:6]: go fmt

File: geography/loader.go [hunk 3]
```diff
@@ -34,9 +41,15 @@ func convertToHierarchy(data *GeographicData) *Hierarchy {
 		entry := NewEntry()
 		entry.Code = code
 		if item.AreaType != nil {
-			entry.Level = int(item.AreaType.Level)
-			entry.Abbreviation = item.AreaType.Abbreviation
-			entry.Codename = item.AreaType.Codename
+			key := item.AreaType.Abbreviation
+			if area, exists := hierarchy.AreaTypes[key]; exists {
+				if (area.Name != item.AreaType.Codename) {
+					fmt.Println("WARNING: AreaType %s is defined multiple times with different names - '%s', '$s'", key, area.Name, item.AreaType.Codename)
```
Comment (NeilMadden) [inline geography/loader.go:49]: Should this be a fatal error?

File: geography/loader_test.go [hunk 4]
```diff
@@ -42,3 +45,20 @@ func TestReadDataToHierarchy(t *testing.T) {
...
 }
+
+func TestReadEmptyDataToHierarchy(t *testing.T) {
+
+	Convey("Given a reader containing nothing", t, func() {
+		readcloser := ioutil.NopCloser(strings.NewReader("{}"))
+
+		Convey("When read into a hierarchy", func() {
+			defer func() {
+				r := recover()
```
Comment (NeilMadden) [inline geography/loader_test.go:56]: Should this be a panic or an `error` result? (I'm not fluent enough in Go to say one way or the other).

Related: if these `So(...)` assertions fail in the middle of a deferred block, does that get reported correctly?

File: sql/exporter.go [hunk 5]
```diff
@@ -4,27 +4,81 @@ import (
...
 	"io"
 	"strings"
+	"sort"
 )
 
 const (
-	hierarchySql = "insert into hierarchy (hierarchy_id, hierarchy_name) values ('%s', %s);\n"
-	entrySql     = "insert into hierarchy_entry (hierarchy_id, entry_code, parent_code, code_name, abbreviation, description, level) values ('%s', '%s', %s, %s, %s, %s, %d);\n"
+	hierarchySql = "insert into hierarchy (hierarchy_id, hierarchy_name) values (%s, %s);\n"
+	areaSql = "insert into hierarchy_area_type (id, name, level) select %s, %s, %d where not exists (select id from hierarchy_area_type where id=%[1]s);\n"
```
Comment (NeilMadden) [inline sql/exporter.go]: Is the `where not exists` clause to prevent inserting duplicates? Could the same be achieve with `insert into ... values ... on conflict do nothing`?

File: sql/exporter.go [hunk 6]
```diff
@@ -4,27 +4,81 @@ import (
...
 // Please note that there is no error handling of failed writes
 // This is a stand-alone command line tool, so errors can just be reported to the user
-func WriteInserts(writer io.Writer, hierarchy *Hierarchy) {
+func WriteSql(writer io.Writer, hierarchy *Hierarchy) {
 
 	if len(hierarchy.Id) == 0 {
 		panic("Cannot write sql for a hierarchy without an id!")
 	}
-	io.WriteString(writer, fmt.Sprintf(hierarchySql, hierarchy.Id, quote(hierarchy.Names["en"])))
+	io.WriteString(writer, fmt.Sprintf(hierarchySql, quote(hierarchy.Id), quote(hierarchy.Names["en"])))
```
Comment (NeilMadden) [inline sql/exporter.go]: This `Sprintf` and manually quoting makes me nervous. Is there not a way we can use prepared statements and bind parameters here?

**Review by NeilMadden — APPROVED** [review]
A few questions about style, but nothing that should block this going in. Feel free to merge when you are happy with it.

File: geography/loader_test.go (hunk 4 above)
Comment (YetAnotherMatt) [reply geography/loader_test.go:56]: The So(...) assertions do get reported correctly, yes.

File: sql/exporter.go (hunk 6 above)
Comment (YetAnotherMatt) [reply sql/exporter.go]: This is writing sql statements to a file, so we can't avoid quotes. There is no actual interaction with a db.

**Review by NeilMadden — APPROVED** [review]
